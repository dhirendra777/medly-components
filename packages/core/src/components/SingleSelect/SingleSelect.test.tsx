import { cleanup, fireEvent, render } from '@test-utils';
import React from 'react';
import { SingleSelect } from './SingleSelect';

describe('SingleSelect component', () => {
    afterEach(cleanup);

    const options = [
        { value: 'all', label: 'All' },
        { value: 'Dummy1', label: 'Dummy1' },
        { value: 'Dummy2', label: 'Dummy2' }
    ];

    it('should render correctly with default props', () => {
        const { container } = render(<SingleSelect options={options} />);
        expect(container).toMatchSnapshot();
    });

    it('should render correctly with all the props given', () => {
        const mockOnChange = jest.fn();
        const { container } = render(
            <SingleSelect
                disabled
                minWidth={300}
                options={options}
                value="all"
                labelPosition="left"
                fullWidth
                required
                label="Pharmacy"
                description="We will show reports based on Pharmacy"
                descriptionColor="grey"
                placeholder="Please Select"
                onChange={mockOnChange}
            />
        );
        expect(container).toMatchSnapshot();
    });

    it('should render options correctly', () => {
        const { container } = render(<SingleSelect options={options} />),
            inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
        fireEvent.click(inputEl);
        expect(container.querySelector('#medly-popover')).toBeVisible();
    });

    it('should hide options on click outside of the container', () => {
        const { container } = render(
            <div>
                <p id="sibling">Outer Element</p>
                <SingleSelect options={options} />
            </div>
        );
        fireEvent.click(container.querySelector('#medly-singleSelect-input'));
        fireEvent.click(container.querySelector('#sibling'));
        expect(container.querySelector('#medly-popover')).toBeNull();
    });

    it('should not show options on click outside of the container', () => {
        const { container } = render(
            <div>
                <p id="sibling">Outer Element</p>
                <SingleSelect options={options} />
            </div>
        );
        fireEvent.click(container.querySelector('#sibling'));
        expect(container.querySelector('#medly-popover')).toBeNull();
    });

    it('should show selected option label in input', () => {
        const mockOnChange = jest.fn(),
            { container, getByText } = render(<SingleSelect options={options} onChange={mockOnChange} />),
            inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;

        fireEvent.click(inputEl);
        fireEvent.click(getByText('Dummy1'));
        expect(inputEl.value).toEqual('Dummy1');
        expect(mockOnChange).toHaveBeenCalledWith('Dummy1');
    });

    it('should not select option which is disable', () => {
        const mockOnChange = jest.fn(),
            { container, getByText } = render(
                <SingleSelect
                    value="Dummy1"
                    options={[...options, { value: 'Dummy3', label: 'Dummy3', disabled: true }]}
                    onChange={mockOnChange}
                />
            );
        const inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
        fireEvent.click(inputEl);
        fireEvent.click(getByText('Dummy3'));
        expect(inputEl.value).toEqual('Dummy1');
        expect(mockOnChange).not.toBeCalled();
    });

    it('should render matched options when input values changes', async () => {
        const { container, queryByText } = render(<SingleSelect options={options} />),
            inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
        fireEvent.click(inputEl);
        fireEvent.change(inputEl, { target: { value: 'Dummy' } });
        expect(queryByText('All')).toBeNull();
    });

    it('should render all the options when input value is not matched to any option', async () => {
        const { container, getByText } = render(<SingleSelect options={options} />),
            inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
        fireEvent.click(inputEl);
        fireEvent.change(inputEl, { target: { value: 'Hello' } });
        expect(getByText('All')).toBeVisible();
        expect(getByText('Dummy1')).toBeVisible();
        expect(getByText('Dummy2')).toBeVisible();
    });

    describe('on pressing', () => {
        afterEach(cleanup);

        describe('down arrow', () => {
            it('should change input', async () => {
                const mockOnChange = jest.fn(),
                    { container, getByText } = render(<SingleSelect value="Dummy1" options={options} onChange={mockOnChange} />),
                    inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
                fireEvent.focus(inputEl);
                fireEvent.keyDown(container, { key: 'ArrowDown', code: 40 });
                fireEvent.keyDown(container, { key: 'Enter', code: 13 });
                expect(inputEl.value).toEqual('Dummy2');
                expect(mockOnChange).toHaveBeenCalledWith('Dummy2');
            });

            it('should change input value to the first option when currently selected options is the last option ', async () => {
                const mockOnChange = jest.fn(),
                    { container } = render(<SingleSelect value="Dummy2" options={options} onChange={mockOnChange} />),
                    inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
                fireEvent.click(inputEl);
                fireEvent.keyDown(container, { key: 'ArrowDown', code: 40 });
                fireEvent.keyDown(container, { key: 'Enter', code: 13 });
                expect(inputEl.value).toEqual('All');
                expect(mockOnChange).toHaveBeenCalledWith('all');
            });

            it('should change input value to the first option when there is no option selected ', async () => {
                const mockOnChange = jest.fn(),
                    { container } = render(<SingleSelect options={options} onChange={mockOnChange} />),
                    inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
                fireEvent.click(inputEl);
                fireEvent.keyDown(container, { key: 'ArrowDown', code: 40 });
                fireEvent.keyDown(container, { key: 'Enter', code: 13 });
                expect(inputEl.value).toEqual('All');
                expect(mockOnChange).toHaveBeenCalledWith('all');
            });

            it('should change input value to the first option when last option is disabled', async () => {
                const mockOnChange = jest.fn(),
                    { container } = render(
                        <SingleSelect
                            value="Dummy2"
                            options={[...options, { value: 'Dummy3', label: 'Dummy3', disabled: true }]}
                            onChange={mockOnChange}
                        />
                    ),
                    inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
                fireEvent.click(inputEl);
                fireEvent.keyDown(container, { key: 'ArrowDown', code: 40 });
                fireEvent.keyDown(container, { key: 'Enter', code: 13 });
                expect(inputEl.value).toEqual('All');
                expect(mockOnChange).toHaveBeenCalledWith('all');
            });

            it('should change input value to the next non-disabled option', async () => {
                const mockOnChange = jest.fn(),
                    { container } = render(
                        <SingleSelect
                            value="Dummy2"
                            options={[
                                ...options,
                                { value: 'Dummy3', label: 'Dummy3', disabled: true },
                                { value: 'Dummy4', label: 'Dummy4' }
                            ]}
                            onChange={mockOnChange}
                        />
                    ),
                    inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
                fireEvent.click(inputEl);
                fireEvent.keyDown(container, { key: 'ArrowDown', code: 40 });
                fireEvent.keyDown(container, { key: 'Enter', code: 13 });
                expect(inputEl.value).toEqual('Dummy4');
                expect(mockOnChange).toHaveBeenCalledWith('Dummy4');
            });
        });

        describe('up arrow', () => {
            it('should change input value', async () => {
                const mockOnChange = jest.fn(),
                    { container } = render(<SingleSelect value="Dummy2" options={options} onChange={mockOnChange} />),
                    inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
                fireEvent.click(inputEl);
                fireEvent.keyDown(container, { key: 'ArrowUp', code: 38 });
                fireEvent.keyDown(container, { key: 'Enter', code: 13 });
                expect(inputEl.value).toEqual('Dummy1');
                expect(mockOnChange).toHaveBeenCalledWith('Dummy1');
            });

            it('should change input value to the last option when currently selected option is the first option', async () => {
                const mockOnChange = jest.fn(),
                    { container } = render(<SingleSelect value="all" options={options} onChange={mockOnChange} />),
                    inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
                fireEvent.click(inputEl);
                fireEvent.keyDown(container, { key: 'ArrowUp', code: 38 });
                fireEvent.keyDown(container, { key: 'Enter', code: 13 });
                expect(inputEl.value).toEqual('Dummy2');
                expect(mockOnChange).toHaveBeenCalledWith('Dummy2');
            });

            it('should change input value to the last option when there is no option selected ', async () => {
                const mockOnChange = jest.fn(),
                    { container } = render(<SingleSelect options={options} onChange={mockOnChange} />),
                    inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
                fireEvent.click(inputEl);
                fireEvent.keyDown(container, { key: 'ArrowUp', code: 38 });
                fireEvent.keyDown(container, { key: 'Enter', code: 13 });
                expect(inputEl.value).toEqual('Dummy2');
                expect(mockOnChange).toHaveBeenCalledWith('Dummy2');
            });

            it('should change input value to the prev non-disabled option', async () => {
                const mockOnChange = jest.fn(),
                    { container } = render(
                        <SingleSelect
                            value="Dummy4"
                            options={[
                                ...options,
                                { value: 'Dummy3', label: 'Dummy3', disabled: true },
                                { value: 'Dummy4', label: 'Dummy4' }
                            ]}
                            onChange={mockOnChange}
                        />
                    ),
                    inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
                fireEvent.click(inputEl);
                fireEvent.keyDown(container, { key: 'ArrowUp', code: 40 });
                fireEvent.keyDown(container, { key: 'Enter', code: 13 });
                expect(inputEl.value).toEqual('Dummy2');
                expect(mockOnChange).toHaveBeenCalledWith('Dummy2');
            });

            it('should change input value to the last option when first option is disabled and second is currently selected', async () => {
                const mockOnChange = jest.fn(),
                    { container } = render(
                        <SingleSelect
                            value="all"
                            options={[{ value: 'Dummy3', label: 'Dummy3', disabled: true }, ...options]}
                            onChange={mockOnChange}
                        />
                    ),
                    inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
                fireEvent.click(inputEl);
                fireEvent.keyDown(container, { key: 'ArrowUp', code: 40 });
                fireEvent.keyDown(container, { key: 'Enter', code: 13 });
                expect(inputEl.value).toEqual('Dummy2');
                expect(mockOnChange).toHaveBeenCalledWith('Dummy2');
            });
        });

        describe('enter button', () => {
            it('should close options', async () => {
                const mockOnChange = jest.fn(),
                    { container } = render(<SingleSelect value="Dummy1" options={options} onChange={mockOnChange} />),
                    inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
                fireEvent.click(inputEl);
                fireEvent.keyDown(container, { key: 'ArrowDown', code: 40 });
                fireEvent.keyDown(container, { key: 'Enter', code: 13 });
                expect(inputEl.value).toEqual('Dummy2');
                expect(mockOnChange).toHaveBeenCalledWith('Dummy2');
            });

            it('should not change input value when option are not visible', async () => {
                const mockOnChange = jest.fn(),
                    { container } = render(<SingleSelect value="All" options={options} onChange={mockOnChange} />),
                    inputEl = container.querySelector('#medly-singleSelect-input') as HTMLInputElement;
                fireEvent.click(inputEl);
                fireEvent.change(inputEl, { target: { value: 'Dummy' } });
                fireEvent.keyDown(container, { key: 'Enter', code: 13 });
                expect(inputEl.value).toEqual('Dummy');
                expect(container.querySelector('#medly-popover')).toBeVisible();
            });
        });
    });
});
