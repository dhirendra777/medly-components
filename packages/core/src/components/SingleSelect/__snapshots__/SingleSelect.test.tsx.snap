// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleSelect component should render correctly with all the props given 1`] = `
.c2 {
  font-size: 14px;
  font-weight: 400;
  color: #616161;
  margin: 0px 8px 0px 0px;
  cursor: default;
}

.c2::after {
  color: red;
  content: ' *';
}

.c12 {
  font-size: 14px;
  font-weight: 400;
  color: grey;
}

.c3 {
  grid-area: label;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c13 {
  grid-area: description;
  margin: 5px 0px 0px 2px;
  color: grey;
}

.c6 {
  margin: 8px 8px 8px 0px;
  display: inline-grid;
  grid-template-areas: 'label' 'field' 'description';
  grid-template-columns: auto;
  grid-template-areas: 'field' 'description';
}

.c6 .c15 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  margin: 8px 0px;
  display: grid;
  min-width: 300px;
  grid-template-columns: minmax(max-content,25%) minmax(auto,80%);
  grid-template-areas: 'label field' 'random description';
}

.c1 .c15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-clip: padding-box;
  box-sizing: border-box;
  border: 1px solid #E0E0E0;
  border-radius: 5px;
  overflow: hidden;
}

.c7:focus-within {
  border-color: #012040;
}

.c7 > * {
  padding: 8px;
  font-size: 14px;
  font-weight: 400;
  color: #616161;
}

.c9 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  outline: 0;
  border: 0;
  cursor: not-allowed;
  background-color: #E0E0E0;
}

.c9::-webkit-input-placeholder {
  color: #9E9E9E;
}

.c9::-moz-placeholder {
  color: #9E9E9E;
}

.c9:-ms-input-placeholder {
  color: #9E9E9E;
}

.c9::placeholder {
  color: #9E9E9E;
}

.c9::-webkit-file-upload-button {
  visibility: hidden;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  -webkit-flex-direction: column-reverse;
  -ms-flex-direction: column-reverse;
  flex-direction: column-reverse;
}

.c4 .c16 {
  top: calc(100% + 5px);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c10 {
  font-size: 16px;
  cursor: pointer;
}

.c10 * {
  fill: #616161;
}

.c14 .c0 {
  margin: 0px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c14 .c0 .c8 {
  margin: 0px;
  border: none;
  background-color: transparent;
}

.c5 {
  border: 1px solid #E0E0E0;
  background-color: #E0E0E0;
  cursor: not-allowed;
  border-radius: 4px;
  box-sizing: border-box;
  overflow: hidden;
  height: 35px;
  min-width: 150px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: calc(100% - 0px);
}

.c5 .c0 {
  margin: 0px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c5 .c0 .c8 {
  margin: 0px;
  border: none;
  background-color: transparent;
}

.c5 * {
  cursor: not-allowed;
}

.c11 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  max-width: 20px;
  margin-right: 8px;
  font-size: 10px;
}

<div>
  <div
    class="c0 c1"
    id="medly-singleSelect-field"
  >
    <label
      class="c2 c3"
      for="medly-singleSelect"
      required=""
    >
      Pharmacy
    </label>
    <div
      class="c4"
      id="medly-popover-wrapper"
    >
      <div
        class="c5"
        disabled=""
      >
        <div
          class="c0 c6"
          id="medly-singleSelect-input-field"
        >
          
          <div
            class="c7"
          >
            <input
              autocomplete="off"
              class="c8 c9"
              disabled=""
              id="medly-singleSelect-input"
              placeholder="Please Select"
              required=""
              type="text"
              value="All"
            />
          </div>
          
        </div>
        <svg
          class="c10 c11"
          height="1em"
          theme="[object Object]"
          viewBox="0 0 12 8"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.88 1.29L6 5.17 2.12 1.29A.996.996 0 1 0 .71 2.7L5.3 7.29c.39.39 1.02.39 1.41 0L11.3 2.7a.996.996 0 0 0 0-1.41c-.39-.38-1.03-.39-1.42 0z"
          />
        </svg>
      </div>
    </div>
    <span
      class="c12 c13"
    >
      We will show reports based on Pharmacy
    </span>
  </div>
</div>
`;

exports[`SingleSelect component should render correctly with default props 1`] = `
.c1 {
  margin: 8px 8px 8px 0px;
  display: inline-grid;
  grid-template-columns: minmax(max-content,25%) minmax(auto,80%);
  grid-template-areas: 'label field' 'random description';
  grid-template-columns: auto;
  grid-template-areas: 'field' 'description';
}

.c1 .c10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c4 {
  margin: 8px 8px 8px 0px;
  display: inline-grid;
  grid-template-areas: 'label' 'field' 'description';
  grid-template-columns: auto;
  grid-template-areas: 'field' 'description';
}

.c4 .c10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-clip: padding-box;
  box-sizing: border-box;
  border: 1px solid #E0E0E0;
  border-radius: 5px;
  overflow: hidden;
}

.c5:focus-within {
  border-color: #012040;
}

.c5 > * {
  padding: 8px;
  font-size: 14px;
  font-weight: 400;
  color: #616161;
}

.c7 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  outline: 0;
  border: 0;
  cursor: inherit;
  background-color: #ffffff;
}

.c7::-webkit-input-placeholder {
  color: #9E9E9E;
}

.c7::-moz-placeholder {
  color: #9E9E9E;
}

.c7:-ms-input-placeholder {
  color: #9E9E9E;
}

.c7::placeholder {
  color: #9E9E9E;
}

.c7::-webkit-file-upload-button {
  visibility: hidden;
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  -webkit-flex-direction: column-reverse;
  -ms-flex-direction: column-reverse;
  flex-direction: column-reverse;
}

.c2 .c11 {
  top: calc(100% + 5px);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c8 {
  font-size: 16px;
  cursor: pointer;
}

.c8 * {
  fill: #616161;
}

.c3 {
  border: 1px solid #E0E0E0;
  background-color: #ffffff;
  cursor: pointer;
  border-radius: 4px;
  box-sizing: border-box;
  overflow: hidden;
  height: 35px;
  min-width: 150px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3 .c0 {
  margin: 0px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 .c0 .c6 {
  margin: 0px;
  border: none;
  background-color: transparent;
}

.c3 * {
  cursor: pointer;
}

.c9 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  max-width: 20px;
  margin-right: 8px;
  font-size: 10px;
}

<div>
  <div
    class="c0 c1"
    id="medly-singleSelect-field"
  >
    
    <div
      class="c2"
      id="medly-popover-wrapper"
    >
      <div
        class="c3"
      >
        <div
          class="c0 c4"
          id="medly-singleSelect-input-field"
        >
          
          <div
            class="c5"
          >
            <input
              autocomplete="off"
              class="c6 c7"
              id="medly-singleSelect-input"
              placeholder="Please Select . . ."
              type="text"
              value=""
            />
          </div>
          
        </div>
        <svg
          class="c8 c9"
          height="1em"
          theme="[object Object]"
          viewBox="0 0 12 8"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.88 1.29L6 5.17 2.12 1.29A.996.996 0 1 0 .71 2.7L5.3 7.29c.39.39 1.02.39 1.41 0L11.3 2.7a.996.996 0 0 0 0-1.41c-.39-.38-1.03-.39-1.42 0z"
          />
        </svg>
      </div>
    </div>
    
  </div>
</div>
`;
